### ISSUE TITLE
Exporting onnx and tensorrt models from pytorch model, the accuracy decreases

### ISSUE THREAD

**Issue by ShihaoHan19980609** [ultralytics/ultralytics#20231]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

deploy code
```
from ultralytics import YOLO

# Load a pretrained YOLO11n model
model = YOLO("train/yolov11_n_b16-100e_AdamW/weights/best.pt")

# model.export(format="onnx")  # onnx
model.export(format="engine")  # tensorrt
```
deploy result
```
/home/han/miniconda3/bin/conda run -n DEIM --no-capture-output python /home/han/Project/ultralytics/deploy.py 
WARNING ⚠️ TensorRT requires GPU export, automatically assigning device=0
Ultralytics 8.3.105 🚀 Python-3.10.16 torch-2.8.0.dev20250413+cu128 CUDA:0 (NVIDIA GeForce RTX 5090 D, 32607MiB)
YOLO11n summary (fused): 100 layers, 2,582,932 parameters, 0 gradients, 6.3 GFLOPs

PyTorch: starting from 'train/yolov11_n_b16-100e_AdamW/weights/best.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 8, 8400) (5.2 MB)

ONNX: starting export with onnx 1.17.0 opset 19...
ONNX: slimming with onnxslim 0.1.50...
ONNX: export success ✅ 0.8s, saved as 'train/yolov11_n_b16-100e_AdamW/weights/best.onnx' (10.1 MB)

TensorRT: starting export with TensorRT 10.9.0.34...
[04/19/2025-21:44:06] [TRT] [I] [MemUsageChange] Init CUDA: CPU -2, GPU +0, now: CPU 915, GPU 2014 (MiB)
[04/19/2025-21:44:12] [TRT] [I] [MemUsageChange] Init builder kernel library: CPU +3102, GPU +486, now: CPU 4168, GPU 2500 (MiB)
[04/19/2025-21:44:12] [TRT] [I] ----------------------------------------------------------------
[04/19/2025-21:44:12] [TRT] [I] Input filename:   train/yolov11_n_b16-100e_AdamW/weights/best.onnx
[04/19/2025-21:44:12] [TRT] [I] ONNX IR version:  0.0.9
[04/19/2025-21:44:12] [TRT] [I] Opset version:    19
[04/19/2025-21:44:12] [TRT] [I] Producer name:    pytorch
[04/19/2025-21:44:12] [TRT] [I] Producer version: 2.8.0
[04/19/2025-21:44:12] [TRT] [I] Domain:           
[04/19/2025-21:44:12] [TRT] [I] Model version:    0
[04/19/2025-21:44:12] [TRT] [I] Doc string:       
[04/19/2025-21:44:12] [TRT] [I] ----------------------------------------------------------------
TensorRT: input "images" with shape(1, 3, 640, 640) DataType.FLOAT
TensorRT: output "output0" with shape(1, 8, 8400) DataType.FLOAT
TensorRT: building FP32 engine as train/yolov11_n_b16-100e_AdamW/weights/best.engine
[04/19/2025-21:44:12] [TRT] [I] Local timing cache in use. Profiling results in this builder pass will not be stored.
[04/19/2025-21:44:48] [TRT] [I] Compiler backend is used during engine build.
[04/19/2025-21:45:44] [TRT] [I] Detected 1 inputs and 1 output network tensors.
[04/19/2025-21:45:45] [TRT] [I] Total Host Persistent Memory: 535632 bytes
[04/19/2025-21:45:45] [TRT] [I] Total Device Persistent Memory: 0 bytes
[04/19/2025-21:45:45] [TRT] [I] Max Scratch Memory: 2764800 bytes
[04/19/2025-21:45:45] [TRT] [I] [BlockAssignment] Started assigning block shifts. This will take 238 steps to complete.
[04/19/2025-21:45:45] [TRT] [I] [BlockAssignment] Algorithm ShiftNTopDown took 6.24746ms to assign 11 blocks to 238 nodes requiring 19252224 bytes.
[04/19/2025-21:45:45] [TRT] [I] Total Activation Memory: 19251200 bytes
[04/19/2025-21:45:45] [TRT] [I] Total Weights Memory: 10457156 bytes
[04/19/2025-21:45:45] [TRT] [I] Compiler backend is used during engine execution.
[04/19/2025-21:45:45] [TRT] [I] Engine generation completed in 93.4121 seconds.
[04/19/2025-21:45:45] [TRT] [I] [MemUsageStats] Peak memory usage of TRT CPU/GPU memory allocators: CPU 1 MiB, GPU 136 MiB
TensorRT: export success ✅ 98.0s, saved as 'train/yolov11_n_b16-100e_AdamW/weights/best.engine' (15.0 MB)

Export complete (101.9s)
Results saved to /home/han/Project/ultralytics/train/yolov11_n_b16-100e_AdamW/weights
Predict:         yolo predict task=detect model=train/yolov11_n_b16-100e_AdamW/weights/best.engine imgsz=640  
Validate:        yolo val task=detect model=train/yolov11_n_b16-100e_AdamW/weights/best.engine imgsz=640 data=/home/han/Project/ultralytics/configs/coco.yaml  
Visualize:       https://netron.app
```

val code
```
    # Load a model
    model = YOLO("train/yolov11_n_b16-100e_AdamW/weights/best.py")
    # model = YOLO("train/yolov11_n_b16-100e_AdamW/weights/best.onnx")
    # model = YOLO("train/yolov11_n_b16-100e_AdamW/weights/best.engine")

    # Validate the model
    model.val(data="/home/han/Project/ultralytics/configs/coco.yaml")  
```
tensorrt val result

```
/home/han/miniconda3/bin/conda run -n DEIM --no-capture-output python /home/han/Project/ultralytics/val.py 
WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.
Ultralytics 8.3.105 🚀 Python-3.10.16 torch-2.8.0.dev20250413+cu128 CUDA:0 (NVIDIA GeForce RTX 5090 D, 32607MiB)
Loading /home/han/Project/ultralytics/train/yolov11_n_b16-100e_AdamW/weights/best.engine for TensorRT inference...
[04/19/2025-21:47:27] [TRT] [I] Loaded engine size: 14 MiB
[04/19/2025-21:47:27] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +0, GPU +19, now: CPU 0, GPU 28 (MiB)
val: Scanning /home/han/Project/datasets/coco/labels.cache... 2370 images, 0 backgrounds, 0 corrupt: 100%|██████████| 2370/2370 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2370/2370 [00:19<00:00, 123.91it/s]
                   all       2370       4979      0.594      0.501      0.537      0.245
                   D00        860       1250      0.555      0.438      0.471      0.203
                   D10        666       1189      0.572      0.386      0.435      0.171
                   D20       1445       1893      0.655      0.643      0.682       0.35
                   D40        417        647      0.593      0.537      0.559      0.256
Speed: 0.4ms preprocess, 1.9ms inference, 0.0ms loss, 1.3ms postprocess per image
Results saved to /home/han/ultralytics/runs/detect/val16
```
pytorch model val result

```
/home/han/miniconda3/bin/conda run -n DEIM --no-capture-output python /home/han/Project/ultralytics/val.py 
Ultralytics 8.3.105 🚀 Python-3.10.16 torch-2.8.0.dev20250413+cu128 CUDA:0 (NVIDIA GeForce RTX 5090 D, 32607MiB)
YOLO11n summary (fused): 100 layers, 2,582,932 parameters, 0 gradients, 6.3 GFLOPs
val: Scanning /home/han/Project/datasets/coco/labels.cache... 2370 images, 0 backgrounds, 0 corrupt: 100%|██████████| 2370/2370 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2370/2370 [00:24<00:00, 96.88it/s]
                   all       2370       4979      0.598       0.53      0.563       0.26
                   D00        860       1250      0.571      0.471      0.503      0.226
                   D10        666       1189      0.575      0.396      0.457       0.18
                   D20       1445       1893      0.665      0.681      0.713      0.373
                   D40        417        647       0.58      0.572      0.579      0.263
Speed: 0.4ms preprocess, 5.0ms inference, 0.0ms loss, 1.3ms postprocess per image
Results saved to /home/han/ultralytics/runs/detect/val18
```

environment：
```
absl-py                  2.2.2
accelerate               1.6.0
albucore                 0.0.24
albumentations           1.4.15
annotated-types          0.7.0
calflops                 0.3.2
certifi                  2025.1.31
charset-normalizer       3.4.1
coloredlogs              15.0.1
contourpy                1.3.1
cycler                   0.12.1
eval_type_backport       0.2.2
faster-coco-eval         1.6.5
filelock                 3.16.1
flatbuffers              25.2.10
fonttools                4.57.0
fsspec                   2024.10.0
grpcio                   1.71.0
huggingface-hub          0.30.2
humanfriendly            10.0
idna                     3.10
imageio                  2.37.0
Jinja2                   3.1.4
kiwisolver               1.4.8
lazy_loader              0.4
Markdown                 3.8
MarkupSafe               2.1.5
matplotlib               3.10.1
mpmath                   1.3.0
narwhals                 1.34.1
networkx                 3.4.2
numpy                    1.26.4
nvidia-cublas-cu12       12.8.3.14
nvidia-cuda-cupti-cu12   12.8.57
nvidia-cuda-nvrtc-cu12   12.8.61
nvidia-cuda-runtime-cu12 12.8.57
nvidia-cudnn-cu12        9.8.0.87
nvidia-cufft-cu12        11.3.3.41
nvidia-cufile-cu12       1.13.0.11
nvidia-curand-cu12       10.3.9.55
nvidia-cusolver-cu12     11.7.2.55
nvidia-cusparse-cu12     12.5.7.53
nvidia-cusparselt-cu12   0.6.3
nvidia-nccl-cu12         2.26.2
nvidia-nvjitlink-cu12    12.8.61
nvidia-nvtx-cu12         12.8.55
onnx                     1.17.0
onnxruntime              1.21.1
onnxruntime-gpu          1.21.0
onnxslim                 0.1.50
opencv-python            4.11.0.86
opencv-python-headless   4.11.0.86
packaging                24.2
pandas                   2.2.3
pillow                   11.2.1
pip                      25.0
plotly                   6.0.1
protobuf                 6.30.2
psutil                   7.0.0
py-cpuinfo               9.0.0
pycocotools              2.0.8
pydantic                 2.11.3
pydantic_core            2.33.1
pyparsing                3.2.3
python-dateutil          2.9.0.post0
pytorch-triton           3.3.0+git96316ce5
pytz                     2025.2
PyYAML                   6.0.2
regex                    2024.11.6
requests                 2.32.3
safetensors              0.5.3
scikit-image             0.25.2
scipy                    1.15.2
seaborn                  0.13.2
setuptools               75.8.0
simsimd                  6.2.1
six                      1.17.0
stringzilla              3.12.4
sympy                    1.13.3
tensorboard              2.19.0
tensorboard-data-server  0.7.2
tensorrt                 10.9.0.34
tensorrt_cu12            10.9.0.34
tensorrt_cu12_bindings   10.9.0.34
tensorrt_cu12_libs       10.9.0.34
tifffile                 2025.3.30
tokenizers               0.20.3
torch                    2.8.0.dev20250413+cu128
torchaudio               2.6.0.dev20250413+cu128
torchvision              0.22.0.dev20250413+cu128
tqdm                     4.67.1
transformers             4.45.2
typing_extensions        4.12.2
typing-inspection        0.4.0
tzdata                   2025.2
ultralytics              8.3.105                  /home/han/Project/ultralytics
ultralytics-thop         2.0.14
urllib3                  2.4.0
Werkzeug                 3.1.3
wheel                    0.45.1
             
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ShihaoHan19980609, thank you for your detailed report and for using Ultralytics 🚀! We appreciate you sharing both your export and validation process.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further. This typically includes a small subset of your data, the exact code used, and the steps needed to reproduce the issue.

If you’re exploring custom training or export workflows, please ensure you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) and include relevant details such as dataset samples and log outputs.

For more help or to discuss with the Ultralytics community:
- Join us on [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chat
- Visit our [Discourse](https://community.ultralytics.com/) forum for in-depth discussions
- Check out our [Subreddit](https://reddit.com/r/Ultralytics) for community Q&A and sharing

## Upgrade

Please make sure you’re running the latest version of `ultralytics` and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in several verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response to help you get started. An Ultralytics engineer will review your issue and assist you soon! 😊

**Comment by Y-T-G** [comment]
If you export with `dynamic=True`, it should be same

**Comment by ShihaoHan19980609** [comment]
> If you export with `dynamic=True`, it should be same

Thank you~ 
I tried it, but it doesn't seem to work.
The result was the same as before.

**Comment by Y-T-G** [comment]
PyTorch uses minimum rectangle during validation.

https://github.com/ultralytics/ultralytics/issues/2783#issuecomment-2413530537

It should be the same if you use `rect=False` for both.

Or if you use dynamic model with this PR:
https://github.com/ultralytics/ultralytics/pull/20232

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in the linked PR #20232, thanks to the efforts of @Y-T-G with contributions from @glenn-jocher.

Key changes in this PR include smarter detection of dynamic models during validation, improved handling of batch sizes, and automatic disabling of rectangular inference (`rect`) for dynamic models—not just for PyTorch models. There are also refinements to data loader worker settings for CPU and Apple MPS devices, which should make validation both faster and more reliable across different hardware and model formats.

To test if these changes resolve the issue, try installing the latest code directly:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Alternatively, you can wait for the next official PyPI release, which will include these updates.

Feedback on whether this resolves the problem would be greatly appreciated. If you continue to experience any related issues, please let us know—thanks again for reporting this and helping improve the project! 🙏
